### PR TITLE
Various cleanup

### PR DESCRIPTION
**Pull request by Ms2ger**

### CODE REVIEW

File: src/libsyntax/ast_map/mod.rs [hunk 1]
```diff
@@ -75,21 +75,8 @@ impl<'a> Iterator for LinkedPath<'a> {
...
-
-    fn next(&mut self) -> Option<T> {
-        let &mut Values(ref mut items) = self;
-        items.next().map(|&x| x)
-    }
-}
-
 /// The type of the iterator used by with_path.
-pub type PathElems<'a, 'b> = iter::Chain<Values<'a, PathElem>, LinkedPath<'b>>;
+pub type PathElems<'a, 'b> = iter::Chain<iter::Cloned<slice::Iter<'a, PathElem>>, LinkedPath<'b>>;
```
Comment (eddyb) [inline src/libsyntax/ast_map/mod.rs]: :scream_cat: This is amazing!

File: src/librustc/middle/check_match.rs [hunk 2]
```diff
@@ -225,12 +225,12 @@ fn check_expr(cx: &mut MatchCheckCtxt, ex: &ast::Expr) {
         },
         ast::ExprForLoop(ref pat, _, _, _) => {
             let mut static_inliner = StaticInliner::new(cx.tcx);
-            is_refutable(cx, &*static_inliner.fold_pat((*pat).clone()), |uncovered_pat| {
+            if let Some(uncovered_pat) = refutable_pattern(cx, &*static_inliner.fold_pat((*pat).clone())) {
```
Comment (eddyb) [inline src/librustc/middle/check_match.rs]: Line longer than 100 chars.

File: src/librustc/middle/check_match.rs [hunk 3]
```diff
@@ -966,25 +966,23 @@ fn check_fn(cx: &mut MatchCheckCtxt,
...
-fn is_refutable<A, F>(cx: &MatchCheckCtxt, pat: &Pat, refutable: F) -> Option<A> where
-    F: FnOnce(&Pat) -> A,
-{
+fn refutable_pattern(cx: &MatchCheckCtxt, pat: &Pat) -> Option<P<Pat>> {
     let pats = Matrix(vec!(vec!(pat)));
     match is_useful(cx, &pats, &[DUMMY_WILD_PAT], ConstructWitness) {
         UsefulWithWitness(pats) => {
             assert_eq!(pats.len(), 1);
-            Some(refutable(&*pats[0]))
+            pats.into_iter().next()
```
Comment (eddyb) [inline src/librustc/middle/check_match.rs]: Could you keep this? The fact that we may create new `P<Pat>` shouldn't be exposed where possible.

**Comment by Manishearth** [comment]
This caused a test failure on the rollup

http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/3333/steps/test/logs/stdio

File: src/libsyntax/ast_util.rs [hunk 4]
```diff
@@ -670,20 +670,13 @@ pub fn path_name_eq(a : &ast::Path, b : &ast::Path) -> bool {
...
-            }
-        }
-        true
-    }
+    a.len() == b.len() &&
+    a.iter().zip(b.iter()).all(|(s, t)| {
+        s.identifier.name == t.identifier.name &&
+        // FIXME #7743: ident -> name problems in lifetime comparison?
+        // can types contain idents?
+        s.parameters != t.parameters
```
Comment (eddyb) [inline src/libsyntax/ast_util.rs]: Oops, I didn't notice the inverted condition here, should be `==`. Good thing we have tests.

**Comment by Manishearth** [comment]
You also have an unused `ast_map` import in src/librustc_trans/back/link.rs`

**Comment by Ms2ger** [comment]
Both fixed, sorry about that.

**Pull request merged**